torch/distributed/_composable/fsdp/_fsdp_collectives.py

```diff
@@ -15,7 +15,13 @@ class AllGatherResult(NamedTuple):
     all_gather_output: torch.Tensor
     all_gather_event: Optional[torch.cuda.Event]
     all_gather_work: Optional[dist.distributed_c10d.Work]
-    all_gather_input_numels: List[int]
+    # For each parameter, the all-gather input dtype for each input
+    param_all_gather_input_dtypes: List[List[torch.dtype]]
+    # For each parameter, the all-gather input numel for each input
+    param_all_gather_input_numels: List[List[int]]
```

Comment:
I guess a super ideal way is that we have sth like tree_flatten for the subclass (maybe we can use tensor_flatten?) to flatten the subclass datas and then derive the param_all_gather_input_numels from the flattened tensor inputs?